provider_config.go

```diff
@@ -0,0 +1,57 @@
+package jwtauth
+
+import (
+	"fmt"
+)
+
+// Provider-specific configuration interfaces
+// All providers must implement the CustomProvider interface, and may implement
+// others as needed.
+
+// ProviderMap is used to map a provider name to its provider type
+var ProviderMap = map[string]CustomProvider{
+	// TODO: remove "empty" provider when actual providers are added
+	"empty": &EmptyProvider{},
+}
+
+// CustomProvider - Any custom provider must implement this interface
+type CustomProvider interface {
+	Initialize(*jwtConfig) error
+	SensitiveKeys() []string
+}
+
+// NewProviderConfig - returns appropriate provider struct if provider_config
+// specified in jwtConfig. The provider map is provider name-to-instance of a
+// CustomProvider.
+func NewProviderConfig(jc *jwtConfig, providerMap map[string]CustomProvider) (CustomProvider, error) {
+	var provider string
+	var ok bool
+	var newCustomProvider CustomProvider
+
+	if len(jc.ProviderConfig) == 0 {
+		return nil, nil
+	}
+	if provider, ok = jc.ProviderConfig["provider"].(string); !ok {
+		return nil, fmt.Errorf("'provider' field not found in provider_config")
+	}
+	newCustomProvider, ok = providerMap[provider]
+	if !ok {
+		return nil, fmt.Errorf("provider %q not found in custom providers", provider)
+	}
```

Comment:
This is the opposite scenario I mentioned in other comments, where I think it's more appropriate to not declare `provider`, `ok`, and `newCustomProvider` at the top here an instead have something like so:

```go 
provider, ok := jc.ProviderConfig["provider"].(string)
if !ok {
  return nil, fmt.Errorf("'provider' field not found in provider_config")
}

newCustomProvider, ok := providerMap[provider]
if !ok {
  return nil, fmt.Errorf("provider %q not found in custom providers", provider)
}
```

The reasoning being both `provider` and `newCustomProvider` are used beyond the error checking. Is there good reason to declare them at the start of the function that perhaps I'm not seeing?

Reply:
I think I see what you mean. I don't know of any concrete benefit to declaring everything at the top, beyond avoiding the redeclaration of `ok`, which doesn't seem worth it. Changed in fe083e8.